### PR TITLE
Enforce coersion during array construction

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

**Comment by ruchim** [comment]
:+1:

File: src/main/scala/wdl4s/values/WdlArray.scala [hunk 1]
```diff
@@ -12,25 +13,26 @@ object WdlArray {
       WdlArray(WdlArrayType(WdlStringType), line.split("\t").toSeq.map(WdlString))
     })
   }
-}
 
-case class WdlArray(wdlType: WdlArrayType, value: Seq[WdlValue]) extends WdlValue with TsvSerializable {
+  def apply(wdlType: WdlArrayType, value: Seq[WdlValue]): WdlArray = {
+    if (wdlType == WdlMaybeEmptyArrayType.EmptyArrayType && value.nonEmpty) {
+      throw new UnsupportedOperationException(s"An Array[Void] must be empty but instead has value: ${value.mkString(", ")}")
```
Comment (mcovarr) [inline src/main/scala/wdl4s/values/WdlArray.scala]: I thought you went with `Nothing` over `Void`

File: src/main/scala/wdl4s/values/WdlArray.scala [hunk 2]
```diff
@@ -12,25 +13,26 @@ object WdlArray {
...
-    throw new UnsupportedOperationException(s"Cannot construct ${wdlType.memberType.toWdlString} array with mixed type: ${value.map(_.wdlType).toSet.mkString(", ")}")
-  }
-  if (wdlType == WdlMaybeEmptyArrayType.EmptyArrayType && value.nonEmpty) {
-    throw new UnsupportedOperationException(s"An Array[Void] must be empty but instead has value: ${value.mkString(", ")}")
-  }
-  if (nonEmpty && value.isEmpty) {
-    throw new UnsupportedOperationException(s"An ${wdlType.toWdlString} contain at least one element")
+    val coercedValue = TryUtil.sequence(value map wdlType.memberType.coerceRawValue)
+    coercedValue match {
+      case Success(coercedArray) => new WdlArray(wdlType, coercedArray) {}
```
Comment (mcovarr) [inline src/main/scala/wdl4s/values/WdlArray.scala]: is the `new` required?

**Comment by mcovarr** [comment]
:+1:

File: src/main/scala/wdl4s/values/WdlArray.scala (hunk 2 above)
Comment (cjllanwarne) [inline src/main/scala/wdl4s/values/WdlArray.scala]: Yes, it makes us use the constructor instead of recursing back into the `apply`

**Pull request merged**